### DISCUSSION TITLE
Can JSBSim be encapsulated into a  function？

### DISCUSSION THREAD

**Discussion by theorynice** [JSBSim-Team/jsbsim#362] [Q&A] [answered]
I want to encapsulated the JSBSim into a function in form of _[states]=func(input)_, in which states are the variable we care and inputs are the control faces and throttle. How can I do?

**Comment by theorynice** [comment, marked as answer]
I have found the jsbsim-gym, it may help me

**Comment by bcoconni** [comment]
This issue has been moved to "Discussions". Now that this new feature is available in GitHub, it has been decided to move issues related to support requests and questions to "Discussions" and restrict "Issues" to bugs reports and feature requests.
This change is only motivated by project organization and will not affect the support provided. We'll gladly continue to answer your questions 😄.

**Comment by theorynice** [comment]
@bcoconni Thank you! Now, I am trying to use the Python version of JSBSim. I have installed the jsbsim wheel by _pip install jsbsim_ and move the engine, aircraft and system folders to my new created python project directory. I use the [https://github.com/JSBSim-Team/jsbsim/blob/master/python/JSBSim.py](url) to test if I have installed successfully. When I set the same argument(i.e. scripts/Short_S23_2.xml  --realtime --nice) in this python version and C++ version(JSBSim.cpp), the results from them are different, showing below:
python version:
![image](https://user-images.githubusercontent.com/39797845/102461792-e0b84c00-4083-11eb-8654-0e3e2fd872b0.png)
c++ version
![image](https://user-images.githubusercontent.com/39797845/102461474-625baa00-4083-11eb-9bac-24990cc5601e.png)
As we can see, the python version failed to trim but the c++ version does successfully. 
Is there anything need to be done in addition to _pip install jsbsim_?

**Reply by bcoconni** [reply]
> I use the https://github.com/JSBSim-Team/jsbsim/blob/master/python/JSBSim.py to test if I have installed successfully.

You don't need to do that since `pip` automatically installs `JSBSim.py` in the executable folder of your Python installation. However since there has not been any change to `JSBSim.py` since the release 1.1.2 the result should be the same in both cases. I'm just mentioning that to save you the trouble of downloading and installing `JSBSim.py` for future releases of JSBSim.
> As we can see, the python version failed to trim but the c++ version does successfully.

Yes, I can reproduce the problem. Basically this is due to a mismatch between the aircraft data (bleeding edge from `master`) and the Python module (release 1.1.2) you are using.

More precisely this is linked to the problem described in issue #360. The code in `FGPropeller` has been slightly modified recently (commit 22f8330e557156cfda7 following discussions in issue #333). Unfortunately it broke the trim solver convergence for `Short_S23_2.xml`. So I have modified the script initial conditions in commit e0f1449ca56bcfd98da to help convergence. Obviously, this breaks convergence for code that do not have the `FGPropeller` modifications (and the Python code from release 1.1.2 doesn't: these modifications will be included in the next release). 
> Is there anything need to be done in addition to pip install jsbsim?

Your first option is to use aircraft data compatible with the Python module you installed with `pip`. You can download them from the [release 1.1.2 files](https://github.com/JSBSim-Team/jsbsim/releases/tag/v1.1.2) (either `Source code (zip)` or `Source code (tar.gz)` at your convenience).

If you plan to use bleeding edge code, then your best option is to not use `pip` and clone the git repository and build JSBSim from scratch (build instructions are available [hre](https://github.com/JSBSim-Team/jsbsim/blob/v1.1.2/doc/DevelopersDocs.md#building-jsbsim))

**Reply by theorynice** [reply]
@bcoconni  Thank you! Follwing your advice, the script mentioned above works sucessfully! When I instantiate jsbsim and loadmodel the aircraft in my project, there are some problems happening.  I can load c172p succesfully but failed to load c310 or f16. My code is:
```python
import jsbsim

import sys,os
import time
root_path='.'
path=os.getcwd()
ResetName="reset00"
AircraftName="c310"
fdm=jsbsim.FGFDMExec(root_path,None)
#fdm.set_aircraft_path('aircraft')
#fdm.set_engine_path('engine')
#fdm.set_systems_path('systems')
fdm.load_model('c310/c310',False)
fdm.load_ic(path+'/aircraft/c310/reset00.xml',False)
fdm.set_dt=0.08
```
the issue is failing to open c310ap:
![image](https://user-images.githubusercontent.com/39797845/102563629-aa76dd00-4114-11eb-8121-9ae2773feb13.png)
But when I use the JSBSim.py to run the scripts/c3105.xml, it runs successfully. I believe it's my load model method having something wrong. But I cant jump into the fdm.loadmodel method to find reasons in python moduel.How can I look into the python source code?
Is there a tutorial to show how to use the python moduel?
Thanks very much, the jsbsim team helps me very much during my learning.Thanks.

**Reply by bcoconni** [reply]
There are a few details that must be fixed:
* When calling `load_model()` you must not specify the path. Instead you just supply the plain aircraft name: `c310`
* When calling `load_ic()` you can save the trouble of specifying the path, if you set the second argument to `True`.
* The time step should be specified with `set_dt()`

The script should then look like:
```python
import jsbsim

fdm = jsbsim.FGFDMExec('.') # The path supplied to FGFDMExec is the location of the folders "aircraft", "engines" and "systems"

fdm.load_model('c310') # Load the aircraft c310
fdm.load_ic('reset00.xml', True) # Load the initial conditions specified in aircraft/c310/reset00.xml
fdm.set_dt(0.08) # Set the time step to 0.08
fdm.run_ic() # Initialize the aircraft with initial conditions read from reset00.xml

# Run the model the first 5 simulated seconds
while fdm['simulation/sim-time-sec] < 5.:
    fdm.run()
```

**Reply by theorynice** [reply]
Well done, thanks for your help! Now, I can use the jsbsim module to do some reinforcement learning projects. While I calculate the error between the current position z and the reference position z, I find the error so big. The reason I find is that I use the "position/h-sl-ft" and "position/eci-z-ft "to represent them respectively. From my view, they both represent the altitude of aircraft, why they are so different. Is this because the eci position has added the radius of the earth? The simulate result is showing below:
![image](https://user-images.githubusercontent.com/39797845/104160138-93818d00-542b-11eb-84ab-465e475ce75b.png)
The other question is how do I close the simulator notify while the simulator is running such as:
![image](https://user-images.githubusercontent.com/39797845/104160459-26222c00-542c-11eb-9415-198aa3ecd872.png)
the aircraft I load is f16.
Thank you very much. Expected for your help!

**Comment by seanmcleod** [comment]
@theorynice rather than pasting screenshots of your console output rather cut and paste the text from the console window and enter it using ` ```sh text ``` `, e.g.

```sh
(Pdb) self.fdm["position/h-agl-ft"]
3.137191885329336
```

In terms of the large difference between `h-sl-ft` and `eci-z-ft`  the clue is in the name 😉 ECI - Earth Centered Inertial.

To quit the simulation while it's running simply hit `ctrl-c` in the console window.

**Reply by theorynice** [reply]
Thank you for your answer and I will follow your suggestion. The second question above doesn't mean to quit the simulation. What I want is just to clear the notify appearing in the console. In other words, the jsbsim(python version) simulator just run background without any information showing in the console. The picture shown above is the information automatically appearing when I run the f16 at some times. It does not always appear and I am curious about what triggers it.

**Reply by bcoconni** [reply]
The amount of information logged in the console depends on the static variable `FGJSBBase::debug_lvl`. It can be modified with the environment variable `JSBSIM_DEBUG`.
https://github.com/JSBSim-Team/jsbsim/blob/14c5a8154491f34bc9e1ee92455ae3d28b1def5f/src/FGFDMExec.cpp#L105-L106
Said otherwise, if you set the environment variable `JSBSIM_DEBUG` to `0` then the amount of information logged in the console will be much reduced. The higher the value of `JSBSIM_DEBUG` the more information will be logged in the console. By default it is set to `1`.

**Reply by theorynice** [reply]
Thank you, the problems are all solved perfectly!
